scripts/staging/fm.dml

```diff
@@ -0,0 +1,91 @@
+#-------------------------------------------------------------
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+# 
+#   http://www.apache.org/licenses/LICENSE-2.0
+# 
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+
+/**
+ * Factorization Machines.
+ */
+
+forward = function(matrix[double] X, matrix[double] w0, matrix[double] W, matrix[double] V)
+    return (matrix[double] y) {
+  /*
+   * Computes the model y. 
+   *
+   * Reference: 
+   *  - Factorization Machines, Steffen Rendle.
+   *
+   * Inputs:
+   *  - X : n examples with d features, of shape (n, d).
+   *  - w0: the global bias, of shape (1,).
+   *  - W : the strength of each feature, of shape (d, 1).
+   *  - V : factorized interaction terms, of shape (d, k).
+   *  
+   * Outputs:
+   *  - y : target vector, of shape (n, 1)
+   */
+   
+   y = w0 + X %*% W + 0.5 * rowSums( (X%*% V)^2 - (X^2 %*% V^2) )
+}
+   
+   
+// TODO : These gradients, from the paper, are the gradients w.r.t. the output y of the  
+// FM model, i.e., dy/dPARAM. We will want to optimize some loss function depending 
+// on the problem, so we'll want to compute dloss/dPARAM. 
```

Comment:
Just to clarify here, we want to compute `dloss/dPARAM`, where `PARAM` is each of the weights or the bias.  The paper gives the equations for `dy/dPARAM`.  The `backward` function will receive`dloss/dy`.  To compute `dloss/dPARAM`, invoke the chain rule and simply multiply them together, i.e., `dloss/dPARAM = dloss/dy * dy/dPARAM`.  So, for example, `dloss/dw0 = 1 * dloss/dy`, where `1` is the `dy/dw0` from the paper.

Also, for cleaner notation in the code, I would simplify the names by dropping the `dloss/` part, and just keep the rest.  I.e., `dloss/dy` -> `dy`, and `dloss/dw0` -> `dw0`.  That way, you also assume the `dloss/` part.